src/FNA3D_Driver_Vulkan.c

```diff
@@ -947,31 +947,44 @@ typedef struct BufferMemoryWrapper
 
 typedef struct VulkanBuffer VulkanBuffer;
 
+typedef struct VulkanTextureBuffer VulkanTextureBuffer;
+
 /* Info about free regions of data within a VulkanTextureBuffer */
 typedef struct VulkanFreeTextureRegion
 {
+	VulkanTextureBuffer *buffer;
 	VkDeviceSize offset;
 	VkDeviceSize size;
+	uint32_t bufferIndex;
+	uint32_t sortedIndex;
 } VulkanFreeTextureRegion;
 
+/* Manages texture memory per memory type */
+typedef struct VulkanTextureSubAllocator
+{
+	VkDeviceSize nextAllocationSize;
+	VulkanTextureBuffer **textureBuffers;
+	uint32_t textureBufferCount;
+
+	/* maintains a sorted list of free regions on owned texture buffers
+	 * from largest to smallest for efficient texture allocation
+	 */
+	VulkanFreeTextureRegion **sortedFreeRegions;
+	uint32_t sortedFreeRegionCount;
+	uint32_t sortedFreeRegionCapacity;
+} VulkanTextureSubAllocator;
+
 /* Manages memory for a single region of allocated texture memory */
-typedef struct VulkanTextureBuffer
+struct VulkanTextureBuffer
```

Comment:
Should this still be a typedef?

Reply:
It has to be forward-declared because of mutual reference.

Reply:
Woops, totally missed that, my mistake!